src/transformer/remove_comments.cpp

```diff
@@ -0,0 +1,107 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include "transformer/remove_comments.hpp"
+
+namespace ddwaf::transformer {
+
+bool remove_comments::transform_impl(lazy_string &str)
+{
+    enum class comment_type { unknown, html, c, eol } type = comment_type::unknown;
+
+    std::size_t read = 0;
+    std::size_t write = 0;
+    while (read < str.length()) {
+        type = comment_type::unknown;
+        while (read < str.length()) {
+            if (str.at(read) == '<' && read + 3 < str.length() && str.at(read + 1) == '!' &&
+                str.at(read + 2) == '-' && str.at(read + 3) == '-') {
+                read += 4;
+                type = comment_type::html;
+                break;
+            }
+
+            if (str.at(read) == '-' && read + 1 < str.length() && str.at(read + 1) == '-') {
+                // Don't bother updating the read index since we'll exit anyway
+                type = comment_type::eol; // SQL comment
+                break;
+            }
+
+            if (str.at(read) == '#') {
+                // Don't bother updating the read index since we'll exit anyway
+                type = comment_type::eol; // Shell comment
+                break;
+            }
+
+            if (str.at(read) == '/' && read + 1 < str.length()) {
+                if (str.at(read + 1) == '*') {
+                    read += 2;
+                    type = comment_type::c;
+                    break;
+                }
+
+                if (str.at(read + 1) == '/') {
+                    // Don't bother updating the read index since we'll exit anyway
+                    type = comment_type::eol; // C/C++ end of line comment
+                    break;
+                }
+            }
+
+            if (str.modified()) {
+                str[write] = str.at(read);
+            }
+
+            ++write;
+            ++read;
+        }
+
+        if (type == comment_type::unknown || type == comment_type::eol) {
+            break;
+        }
+
+        if (!str.modified()) {
+            // Force a copy
+            str.finalize();
```

Comment:
I don't get why this is needed.

Reply:
This is an optimisation for the first iteration, when a string has no comments:
- The previous loop finds comments, but it's also used to write text before/between/after comments.
- If the string hasn't been modified, it isn't written to within that loop.
- Once we reach this point, we know there is a comment, so we force a copy here so that the next iteration enables line 53.

This essentially allows for the check to be done at the same time as the transformation, making this a single-pass transformer.

Reply:
I have simplified this to make it a bit more clear.